Refactoring the data source before unnest

somu-imply opened the pull request:
### Motivation
There are a number of use cases that require “flattening” records when processing. The need for this is exemplified in the existed of MVDs, which basically have a native “flattening” built-in. Other SQL systems can also do similar behaviors, they expect an ARRAY type of data and then you put that into a FLATTEN or UNNEST or other operator and it “explodes” out one new row for every value in the array.

 

We need to implement a similar operator for Druid. This dove-tails a bit with some of the recent work done for handling arrays in Druid. Essentially, the operator would take in an array and then “flatten” the array into N (N=number of elements in the array) rows where each row has one of the values from the array.

### Related Work
The inspiration for using unnest as a data source comes from ([Work with arrays  |  BigQuery  |  Google Cloud](https://cloud.google.com/bigquery/docs/reference/standard-sql/arrays) )  where unnest has been used as a data source. Clickhouse also has a flatten functionality ([Array Functions | ClickHouse Docs](https://clickhouse.com/docs/en/sql-reference/functions/array-functions/#arrayflatten) ) but that does not transform a dataset by adding more rows. Since for our use case, unnest can be used with dateExpand functionality (coming after unnest) we model unnest as a data source similar to Bigquery. 
 
### Methodology
**Refactoring** 

The underlying principle here is an operation on a data source that works on a segment and creates additional rows. Joins have a similar principle where the number of rows can be more than the input table after the join operation. The current framework supports that in the following way:

Having a join data source, alongside a factory and then a specialized wrapper ([JoinableFactoryWrapper.java](https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/segment/join/JoinableFactoryWrapper.java) ) around it which creates a function to transform one segment into another by the notion of a segment function.

Have a separate implementation of segment reference through  [HashJoinSegment.java](https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/segment/join/HashJoinSegment.java) which uses a custom storage adapter to access a cursor to the segment for processing.

The goal here is to move out the creation of the segment map function from outside the wrapper to individual datasources. In cases where the segment map function is not an identity function (in case of join and also for unnest) the segment functions can be created accordingly for each data source. This makes the abstraction generic and readily extendable to other datasources we might create in future (e.g flatten)

The changes here:

1. Creation of a segment map function moved into each data source
2. Code refactored from `JoinableFactoryWrapper` to `JoinDataSource`
3. PlannerContext was updated with an additional `JoinableFactoryWrapper` object during this process
4. Several test cases were updated with the change to the `JoinDataSource`
5. Some Guice injector things were updated as a part of the process, additionally nullables need not be bound to CliRouter using noops now
6. New testcases for broadcast join in MSQ
7. `getCacheKey` method updated to move inside each data source

**New changes** (will be in a followup)

After the new abstraction is in place, the following needs to be done for unnest

- Have a separate UnnestDataSource class that deals with unnest
- This should have it’s own segment map function
- A separate UnnestSegment with a custom storage adapter to gain access to cursor in each segment
- Design of the native query for Unnest considering we also need to  pass the column(s) to unnest
- Finally, the SQL query that maps to the native query

Each of these development phases should be backed by unit tests/ integration tests as applicable

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

somu-imply commented:
Travis is failing from a branch coverage from the changes to `BaseLeafFrameProcessor`. While I am working on that I am setting this PR up for review now.

somu-imply commented:
Guice issues in Integration tests. Taking a look

imply-cheddar commented on processing/src/main/java/org/apache/druid/segment/join/JoinableFactoryWrapper.java:
```diff
@@ -340,12 +192,60 @@ static JoinClauseToFilterConversion convertJoinToFilter(
     return new JoinClauseToFilterConversion(null, false);
   }
 
+  public JoinableFactory getJoinableFactory()
+  {
+    return joinableFactory;
+  }
+
+  /**
+   * Compute a cache key prefix for a join data source. This includes the data sources that participate in the RHS of a
+   * join as well as any query specific constructs associated with join data source such as base table filter. This key prefix
+   * can be used in segment level cache or result level cache. The function can return following wrapped in an
```
A bunch of this code looks like it changed, but I think it just had methods re-ordered?  Was that intentional?

somu-imply replied:
The IDE did this. I am using the druid intellij code formatter. Let me check how this can be avoided

imply-cheddar commented on processing/src/main/java/org/apache/druid/segment/join/JoinableFactoryWrapper.java:
```diff
@@ -340,12 +192,60 @@ static JoinClauseToFilterConversion convertJoinToFilter(
+   * @return the optional cache key to be used as part of query cache key
+   * @throws {@link IAE} if this operation is called on a non-join data source
+   */
+  public Optional<byte[]> computeJoinDataSourceCacheKey(
```
The existence of this method is another point where things seem a little suspect.  I think the `DataSource` object needs a `public byte[] getCacheKey()` method.  Most of the current implementations would just return `new byte[]{}`, but the join one should return this (or null if not cacheable).  The unnest implementation will need to override the method as well.

somu-imply replied:
This has been moved. The test cases in `JoinableFactoryWrapperTest` have been refactored and 
moved inside `JoinDataSourceTest`

imply-cheddar commented on server/src/main/java/org/apache/druid/server/LocalQuerySegmentWalker.java:
```diff
@@ -94,12 +93,12 @@ public <T> QueryRunner<T> getQueryRunnerForIntervals(final Query<T> query, final
-        cpuAccumulator,
-        analysis.getBaseQuery().orElse(query)
-    );
+    Function<SegmentReference, SegmentReference> segmentMapFn = analysis.getDataSource()
+                                                                        .createSegmentMapFunction(
+                                                                            query,
+                                                                            cpuAccumulator
+                                                                        );
+
```
Now that there's a lot less arguments here, I'm pretty sure this formatting is no longer the best way to represent this code.  Please condense things.

somu-imply replied:
Done

imply-cheddar reviewed (changes requested):
Some comments/requests for changes. 

Btw, in doing the PR, I came across methods in CacheUtils that are written like

```
  public static <T> boolean isUseSegmentCache(
      Query<T> query,
      @Nullable CacheStrategy<T, Object, Query<T>> cacheStrategy,
      CacheConfig cacheConfig,
      ServerType serverType
  )
  {
    return isQueryCacheable(query, cacheStrategy, cacheConfig, serverType)
           && QueryContexts.isUseCache(query)
           && cacheConfig.isUseCache();
  }
```

`isPopulateCache` is the same.  This is really unfortunate because `cacheConfig.isUseCache` is the cheapest of the calls, `QueryContexts.isUseCache(query)` is the second cheapest and `isQueryCacheable(...)` is the most expensive.  AND logic will fail on the first false and not do other calls, so you always want the cheapest ones first, while we are doing the most expensive operation first.  Can you take a few moments to invert the order of the parameters of the AND as well?

somu-imply commented on processing/src/main/java/org/apache/druid/query/InlineDataSource.java:
```diff
@@ -102,6 +105,55 @@ public static InlineDataSource fromIterable(
+   * from materializing things needlessly. Useful for unit tests that want to compare equality of different
+   * InlineDataSource instances.
+   */
+  private static boolean rowsEqual(final Iterable<Object[]> rowsA, final Iterable<Object[]> rowsB)
```
This was my IDE doing stuff. I reformatted the code in the IDE but it does not seem to push public methods in the top of the file

somu-imply commented on extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/BaseLeafFrameProcessor.java:
```diff
@@ -152,13 +153,23 @@ private boolean initializeSegmentMapFn(final IntSet readableInputs)
+          // was using the JoinableFactoryWrapper to create segment map function.
+          // After refactoring, the segment map function creation is moved to data source
+          // Hence for InputNumberDataSource we are setting the broadcast join helper for the data source
+          // and moving the segment map function creation there
```
Done

somu-imply commented on processing/src/main/java/org/apache/druid/query/JoinDataSource.java:
```diff
@@ -127,12 +155,22 @@ public static JoinDataSource create(
       final JoinType joinType,
-      final DimFilter leftFilter
+      final DimFilter leftFilter,
+      @Nullable @JacksonInject final JoinableFactoryWrapper joinableFactoryWrapper
```
Excellent point ! Removed

somu-imply commented on processing/src/main/java/org/apache/druid/query/Queries.java:
```diff
@@ -191,11 +191,19 @@ public static <T> Query<T> withBaseDataSource(final Query<T> query, final DataSo
+     * We should likely look into moving this functionality into the DataSource object itself so that they
+     * can walk and create new objects on their own.  This will be necessary as we expand the set of DataSources
+     * that do actual work, as each of them will need to show up in this if/then waterfall.
+     */
```
@cheddar this was moved to the `DataSource` , no more `instanceof` checks here

imply-cheddar commented on processing/src/main/java/org/apache/druid/segment/join/JoinableFactoryWrapper.java:
```diff
@@ -77,144 +55,13 @@ public JoinableFactoryWrapper(final JoinableFactory joinableFactory)
+   * <p>
    * See {@link #convertJoinToFilter} for details on the logic.
    */
   @VisibleForTesting
```
Don't really need this annotation on a public method :)

imply-cheddar commented on services/src/main/java/org/apache/druid/cli/CliRouter.java:
```diff
@@ -112,7 +114,8 @@ protected List<? extends Module> getModules()
                 .in(LazySingleton.class);
-
+          
+          binder.bind(JoinableFactory.class).to(NoopJoinableFactory.class).in(LazySingleton.class);
```
This shouldn't be necessary anymore, right?

somu-imply replied:
removed

imply-cheddar reviewed (commented):
Getting very close, I think.

One thing that we need to go back and double check is that a lot of the call sites for the `makeSegmentMapFn` are just passing along a random `AtomicLong` for the cpu time tracker thingie.  That's gonna mess up CPU tracking, so we need to look at all of those sites and give them the real tracker thingie instead of an object that just gets thrown away after things are done.

somu-imply commented on benchmarks/src/test/java/org/apache/druid/benchmark/query/SqlBenchmark.java:
```diff
@@ -88,8 +86,8 @@
-@Warmup(iterations = 5)
-@Measurement(iterations = 15)
+@Warmup(iterations = 3)
+@Measurement(iterations = 10)
```
This was reverted. Wherever the cpuAccumulators were reused, we passed a variable but where they were not used downstream I have used a `new AtomicLong()`.

somu-imply commented on extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/InputNumberDataSource.java:
```diff
@@ -81,6 +92,29 @@ public boolean isConcrete()
+      AtomicLong cpuTime
+  )
+  {
+    final DataSource dataSourceWithInlinedChannelData = broadcastJoinHelper.inlineChannelData(query.getDataSource());
```
I added the check

somu-imply commented on core/src/main/java/org/apache/druid/math/expr/ExprMacroTable.java:
```diff
@@ -44,6 +44,7 @@
 {
   private static final ExprMacroTable NIL = new ExprMacroTable(Collections.emptyList());
 
+
```
aah new line. addressed it

imply-cheddar commented on processing/src/main/java/org/apache/druid/query/InlineDataSource.java:
```diff
@@ -252,7 +252,7 @@ public DataSource withUpdatedDataSource(DataSource newSource)
   public byte[] getCacheKey()
   {
-    return null;
+    return new byte[0];
```
Are you sure that this was cached previously?  It seems like an `InlineDataSource` would either have to include all of the data in the cache key or cause caching to never happen...

imply-cheddar commented on extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/InputNumberDataSource.java:
```diff
@@ -112,7 +119,7 @@ public DataSource withUpdatedDataSource(DataSource newSource)
   public byte[] getCacheKey()
   {
-    return null;
+    return new byte[0];
```
This seems uncacheable to me.

imply-cheddar commented on processing/src/main/java/org/apache/druid/query/QueryDataSource.java:
```diff
@@ -109,7 +109,7 @@ public DataSource withUpdatedDataSource(DataSource newSource)
   public byte[] getCacheKey()
   {
-    return null;
+    return new byte[0];
```
So, the query data source itself is not likely cacheable.  The sub-query that is run from inside the data source is perhaps cacheable, but not the datasource.  Well, unless it's a result level cache, that could work.  But, I'm like 80% that `null` is actually correct here.

imply-cheddar commented on sql/src/main/java/org/apache/druid/sql/calcite/external/ExternalDataSource.java:
```diff
@@ -141,7 +141,7 @@ public DataSource withUpdatedDataSource(DataSource newSource)
   public byte[] getCacheKey()
   {
-    return null;
+    return new byte[0];
```
This should either not be cacheable or it needs to build up an identifier for the external data source...  `null` seems correct for now.

somu-imply replied:
So where the `isCacheable` for the data source is false, the `getCacheKey()` should return null for those data sources. Is my understanding correct ?

imply-cheddar replied:
Basically, yes.  I had forgotten about that method.  In some future, we should refactor to completely eliminate the method...  But let's not take that on yet.

somu-imply commented on extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/BaseLeafFrameProcessor.java:
```diff
@@ -146,19 +147,31 @@ protected SegmentReference mapSegment(final Segment segment)
+        }
+        return retVal;
+      } else {
+        segmentMapFn = Function.identity();
```
To continue with Abhishek's point if the parent datasource is a `JoinDataSource` with child data sources being `InputNumberDataSource`, this would basically set the map function to an identity function which is not right. One way is to first construct the data source using the broadcastJoinHelper and call createSegmentMapFunction of that. I'll discuss this with @imply-cheddar today. I believe the else part can be simplified as
```
else {
      final boolean retVal = broadcastJoinHelper.buildBroadcastTablesIncrementally(readableInputs);
      if (retVal) {
        final DataSource dataSourceWithInlinedChannelData = broadcastJoinHelper.inlineChannelData(query.getDataSource());
        segmentMapFn = dataSourceWithInlinedChannelData.createSegmentMapFunction(query, cpuAccumulator);
      }
      else {
        segmentMapFn = Function.identity();
      }
      return true;
```

imply-cheddar replied:
Is there a test for that case?

somu-imply replied:
Yes there is a `MSQSelectTest.joinTest`

imply-cheddar replied:
If it works before the code change, then why do we need the code change?

imply-cheddar replied:
What I'm trying to say is, if there's not a test that can exercise the corner that is being mentioned, then the corner doesn't exist.  If the corner does exist, there should be a test that exercises it which is broken.  So, we are in 1 of 3 states:

1) There should be a test that's breaking, but there isn't yet.  In that case, let's create the test and then talk about solutions.
2) There is a test that exercises the corner, but it's passing.  In this case, the corner doesn't actually exist and we do not need further code changes.
3) There is no test that generates the highlighted state because something upstream from this code is actually preventing that state from ever existing.  So, it is impossible to actually get to the bad state.  In this case, the corner doesn't actually exist, no code changes required.

somu-imply replied:
This follows case 1. The new test case has been added

clintropolis commented on sql/src/main/java/org/apache/druid/sql/calcite/planner/PlannerContext.java:
```diff
@@ -77,6 +78,7 @@
   private final String sql;
   private final DruidOperatorTable operatorTable;
   private final ExprMacroTable macroTable;
+  private final JoinableFactoryWrapper jfw;
```
super super nit: tiny variable name doesn't seem very consistent with other names 😅

somu-imply replied:
My bad a new round is coming with all these changes (have them locally)

clintropolis commented on server/src/main/java/org/apache/druid/server/LocalQuerySegmentWalker.java:
```diff
@@ -94,12 +93,13 @@ public <T> QueryRunner<T> getQueryRunnerForIntervals(final Query<T> query, final
-        cpuAccumulator,
-        analysis.getBaseQuery().orElse(query)
-    );
+    Function<SegmentReference, SegmentReference> segmentMapFn = analysis
```
super nit: `final` to match other lines 😅

clintropolis commented on processing/src/main/java/org/apache/druid/query/DataSource.java:
```diff
@@ -83,8 +89,40 @@
+   * @return the cache key to be used as part of query cache key
+   * @throws {@link IAE} if this operation is called on a non-join data source
+   */
+  byte[] getCacheKey();
```
>throws {@link IAE} if this operation is called on a non-join data source
the javadocs are maybe not true exactly? it looks like most implementations are currently returning empty bytes.

Im also a bit tunsure how great this naming is given that it should only be used for joins afaik, and not used at all for cache keys otherwise. it seems sort of seems like it would be important for all query types, but calling it something specific to joins also seems short sighted... Not sure what exactly to do here.. maybe this needs a wider discussion about what kinds of datasources need to implement this and what do not which would help clear this up

imply-cheddar replied:
Yeah, that javadoc on the `throws` is old.  It should never actually throw.

At this point, this is only for joins, the semantics that we introduced, though, give the different datasources the ability to completely opt-out of caching if a meaningful cache key cannot be created.  The refactor is actually making this explicit and the various implementations are now opting in/out of caching on their own (by either returning `null` or `new byte[]{}`)

clintropolis reviewed (commented):
overall refactor seems nice :+1:

imply-cheddar commented on extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/InputNumberDataSource.java:
```diff
@@ -81,6 +92,36 @@ public boolean isConcrete()
+    }
+    final DataSource dataSourceWithInlinedChannelData = broadcastJoinHelper.inlineChannelData(query.getDataSource());
+    final DataSourceAnalysis analysis = DataSourceAnalysis.forDataSource(dataSourceWithInlinedChannelData);
+    return analysis.getDataSource().createSegmentMapFunction(query, new AtomicLong());
```
Pass through the cpuTime that was passed into this method please.

somu-imply commented on extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/BaseLeafFrameProcessor.java:
```diff
@@ -146,18 +147,30 @@ protected SegmentReference mapSegment(final Segment segment)
+          inputNumberDataSource.setBroadcastJoinHelper(broadcastJoinHelper);
+          segmentMapFn = inputNumberDataSource.createSegmentMapFunction(query, cpuAccumulator);
+        } else {
+          segmentMapFn = inlineChannelDataSource.createSegmentMapFunction(query, cpuAccumulator);
```
@abhishekagarwal87 I have updated this part. Should address your earlier concern

imply-cheddar replied:
Did we find a test that was broken with the previous code?  If not, then please revert back to the code that @abhishekagarwal87 initially commented on and keep that code active.  That code is not broken until we can actually reproduce the breakage.

somu-imply replied:
This example passes with the new code changes. The problem discovered was :
1. There was a lack of a proper test case for this to be discovered. I am adding the test suggested here 
2. In the previous case, the identity function in the last else part was incorrect as it would not create the correct segment function for this example. This has been now updated by calling `createSegmentMapFunction` on the correct data source.
3. All unit tests for MSQ, CalciteJoins now complete successfully

imply-cheddar commented on server/src/main/java/org/apache/druid/server/coordination/ServerManager.java:
```diff
@@ -195,18 +194,10 @@ public <T> QueryRunner<T> getQueryRunnerForSegments(Query<T> query, Iterable<Seg
     } else {
       return new ReportTimelineMissingSegmentQueryRunner<>(Lists.newArrayList(specs));
     }
+    Function<SegmentReference, SegmentReference> segmentMapFn = analysis.getDataSource().createSegmentMapFunction(query, cpuTimeAccumulator);
```
does this need to be the `analysis.getDataSource()` instead of `query.getDataSource()`?

somu-imply replied:
I get confused between use of `analysis.getDataSource()` and `query.getDataSource()` especially when analysis is created from `DataSourceAnalysis.fromDataSource(query.getDataSource())`. This is the case here. I have updated it to use query

imply-cheddar replied:
Yes it is super easy to get confused by that...

somu-imply commented:
@imply-cheddar , @clintropolis I have addressed the comments and also resolved the conflicts. The thing remaining to address it use of a set in `GuiceInjectableValues` so the nullable check can be avoided

somu-imply commented:
@clintropolis @imply-cheddar @abhishekagarwal87 this is ready for review. Can you please take a look

somu-imply commented:
The conflicts have been resolved. Some additional refactoring introduced due to change in underlying test framework in #12965

imply-cheddar commented on core/src/main/java/org/apache/druid/guice/GuiceInjectableValues.java:
```diff
@@ -50,7 +59,21 @@ public Object findInjectableValue(
+      }
+      catch (ConfigurationException ce) {
+        // check if nullable annotation is present for this
+        if (nullables.get().contains((Key) valueId)) {
```
The point of caching these is so that we only have Guice generate the exception once.  Putting this check inside of the catch means that we don't get any benefit.  This check needs to be done before we even ask the injector to create the instance.

imply-cheddar commented on core/src/main/java/org/apache/druid/guice/GuiceInjectableValues.java:
```diff
@@ -50,7 +59,21 @@ public Object findInjectableValue(
+          return null;
+        } else if (forProperty.getAnnotation(Nullable.class) != null) {
+          HashSet<Key> encounteredNullables = nullables.get();
+          encounteredNullables.add((Key) valueId);
```
This is a non-thread-safe mutation of the HashSet.  The goal of the AtomicReference was to make thinds thread-safe.  You must create a brand new Set, add all of the old values, add the new value and then set the new reference on the AtomicReference.

somu-imply replied:
Addressed
